File: web/src/api/ddns.ts
import client from "./client";

export type DDNSConfig = {
  id: number;
  provider: string;
  accessKeyId: string;
  domain: string;
  subdomain: string;
  enabled: boolean;
  checkIntervalSeconds: number;
  lastKnownIpv6: string;
  lastStatus: string;
  lastError: string;
  lastSyncedAt?: string;
};

export type DDNSPayload = {
  provider: string;
  accessKeyId: string;
  accessKeySecret: string;
  domain: string;
  subdomain: string;
  enabled: boolean;
  checkIntervalSeconds: number;
};

export type DDNSRuntimeStatus = {
  lastKnownIpv6: string;
  lastStatus: string;
  lastError: string;
  lastSyncedAt?: string;
};

export async function fetchDDNSConfigs() {
  const response = await client.get("/ddns");
  return response.data as { items: DDNSConfig[]; runtimeStatus: DDNSRuntimeStatus };
}

export async function createDDNSConfig(payload: DDNSPayload) {
  const response = await client.post("/ddns", payload);
  return response.data as DDNSConfig;
}

export async function updateDDNSConfig(id: number, payload: DDNSPayload) {
  const response = await client.put(`/ddns/${id}`, payload);
  return response.data as DDNSConfig;
}

export async function syncDDNSConfigs() {
  const response = await client.post("/ddns/sync");
  return response.data as { message: string };
}

File: internal/domain/ddns.go
package domain

import "time"

type DDNSConfig struct {
	ID                   int64      `json:"id"`
	Provider             string     `json:"provider"`
	AccessKeyID          string     `json:"accessKeyId"`
	AccessKeySecretEnc   string     `json:"-"`
	Domain               string     `json:"domain"`
	Subdomain            string     `json:"subdomain"`
	Enabled              bool       `json:"enabled"`
	CheckIntervalSeconds int        `json:"checkIntervalSeconds"`
	LastKnownIPv6        string     `json:"lastKnownIpv6"`
	LastStatus           string     `json:"lastStatus"`
	LastError            string     `json:"lastError"`
	LastSyncedAt         *time.Time `json:"lastSyncedAt,omitempty"`
	CreatedAt            time.Time  `json:"createdAt"`
	UpdatedAt            time.Time  `json:"updatedAt"`
}

type DDNSRuntimeStatus struct {
	LastKnownIPv6 string     `json:"lastKnownIpv6"`
	LastStatus    string     `json:"lastStatus"`
	LastError     string     `json:"lastError"`
	LastSyncedAt  *time.Time `json:"lastSyncedAt,omitempty"`
}
